nixos/modules/services/mail/cyrus-imap.nix

```
@@ -0,0 +1,366 @@
+{
+  pkgs,
+  lib,
+  config,
+  ...
+}:
+let
+  cfg = config.services.cyrus-imap;
+  cyrus-imapdPkg = pkgs.cyrus-imapd;
+  inherit (lib)
+    literalExpression
+    mkEnableOption
+    mkIf
+    mkOption
+    optionalAttrs
+    generators
+    mapAttrsToList
+    ;
+  inherit (lib.strings) concatStringsSep;
+  inherit (lib.types)
+    attrsOf
+    submodule
+    listOf
+    oneOf
+    str
+    int
+    bool
+    ints
+    enum
+    nullOr
+    path
+    ;
+
+  mkCyrusConfig =
+    settings:
+    concatStringsSep "\n  " (
+      mapAttrsToList (n: v: v) (
+        builtins.mapAttrs (
+          name: value:
+          concatStringsSep " " (
+            [ "${name}" ]
+            ++ (mapAttrsToList (
+              n: v:
+              if (v != null) then
+                if builtins.isInt v then
+                  "${n}=${builtins.toString v}"
+                else
+                  "${n}=\"${if builtins.isList v then (concatStringsSep " " v) else v}\""
+              else
+                ""
+            ) value)
+          )
+        ) settings
+      )
+    );
+  cyrusConfig = ''
+    START {
+      ${mkCyrusConfig cfg.cyrusSettings.START}
+    }
+    SERVICES {
+      ${mkCyrusConfig cfg.cyrusSettings.SERVICES}
+    }
+    EVENTS {
+      ${mkCyrusConfig cfg.cyrusSettings.EVENTS}
+    }
+    DAEMON {
+      ${mkCyrusConfig cfg.cyrusSettings.DAEMON}
+    }
+  '';
+
+  imapdConfig =
+    with generators;
+    toKeyValue {
+      mkKeyValue = mkKeyValueDefault {
+        mkValueString =
+          v:
+          if builtins.isBool v then
+            if v then "yes" else "no"
+          else if builtins.isList v then
+            concatStringsSep " " v
+          else
+            mkValueStringDefault { } v;
+      } ": ";
+      listsAsDuplicateKeys = false;
+    } cfg.imapdSettings;
+in
+{
+  options.services.cyrus-imap = {
+    enable = mkEnableOption ("Cyrus IMAP, an email, contacts and calendar server");
+
+    debug = mkEnableOption ("enable debugging on cyrus master");
+    listenQueue = mkOption {
+      type = int;
+      default = 32;
+      description = ''
+        Socket listen queue backlog size
+        See listen(2). Default is 32, you may want to increase this number if you have a very high connection rate
+      '';
+    };
+    tmpDBDir = mkOption {
+      type = path;
+      default = "/run/cyrus/db";
+      description = ''
+        Locations for DB files.
+        DBs under this directory are recreated upon initialization, so should live in ephemeral storage for best performance.
+      '';
+    };
+    cyrusSettings = mkOption {
```

Just call this attribute `settings` for consistency with other modules.
```suggestion
    settings = mkOption {
```